rlw/upgrade.ps1

```diff
@@ -0,0 +1,167 @@
+$json = & "${RIGHTLINK_DIR}\rsc.exe" rl10 index /rll/proc
+
+# Determine current version of rightlink
+$currentVersion = Write-Output $json | & "${RIGHTLINK_DIR}\rsc.exe" --x1 .version json 2> $null
+
+if (!$currentVersion) {
+  Write-Output 'Cannot determine current version of RightLink'
+  Exit 1
+}
+
+# Fetch information about what we should become. The "upgrades" file consists of lines formatted
+# as "currentVersion:desiredVersion". If the "upgrades" file does not exist,
+# or if the current version is not in the file, no upgrade is done.
+$upgradesFile = "${PSScriptRoot}\upgrades"
+(New-Object System.Net.WebClient).DownloadFile($env:UPGRADES_FILE_LOCATION, $upgradesFile)
+$desiredVersion = Get-Content $upgradesFile | Select-String "^\s*${currentVersion}\s*:\s*(\S+)\s*$" |
+                  % { $_ -replace "^\s*${currentVersion}\s*:\s*", '' }
+
+if (!$desiredVersion) {
+  Write-Output 'Cannot determine latest version from upgrade file'
+  Write-Output "Tried to match /^${currentVersion}:/ in ${env:UPGRADES_FILE_LOCATION}"
+  Exit 0
+}
+
+if ($desiredVersion -eq $currentVersion) {
+  Write-Output "RightLink is already up-to-date (current = ${currentVersion})"
+  Exit 0
+}
+
+Write-Output 'RightLink needs update:'
+Write-Output "  from current = ${currentVersion}"
+Write-Output "  to   desired = ${desiredVersion}"
+
+Write-Output "Downloading RightLink version '${desiredVersion}'"
+
+# Download new version
+$TMP_DIR = 'C:\Windows\Temp\Upgrade'
+$RIGHTLINK_URL = "https://rightlink.rightscale.com/rll/${desiredVersion}/rightlink.zip"
+
+# If temp directory doesn't exist then create it
+if (!(Test-Path -Path $TMP_DIR)) {
+  New-Item -Path $TMP_DIR -Type Directory | Out-Null
+}
+
+if (Test-Path -Path ${TMP_DIR}\rightlink) {
+  Remove-Item "${TMP_DIR}\rightlink" -Force -Recurse
+}
+if (Test-Path -Path ${TMP_DIR}\rightlink.zip) {
+  Remove-Item "${TMP_DIR}\rightlink.zip" -Force
+}
+
+$wc = New-Object System.Net.WebClient
+$wc.DownloadFile($RIGHTLINK_URL, "${TMP_DIR}\rightlink.zip")
+
+# Expand the archive into C:\Temp\Upgrade
+Expand-Zipfile "${TMP_DIR}\rightlink.zip" $TMP_DIR | Out-Null
+
+# Check downloaded version
+if (Test-Path -Path ${RIGHTLINK_DIR}\rightlink-new.exe) {
+  Remove-Item "${RIGHTLINK_DIR}\rightlink-new.exe" -Force
+}
+Move-Item -Path "${TMP_DIR}\RightLink\rightlink.exe" -Destination "${RIGHTLINK_DIR}\rightlink-new.exe" -Force
+Write-Output 'Checking new version'
+$newVersion = & "${RIGHTLINK_DIR}\rightlink-new.exe" --version | % { $_.Split(" ")[1] }
+
+if ($newVersion -eq $desiredVersion) {
+  Write-Output "New version looks right: ${newVersion}"
+  Write-Output 'Restarting RightLink to pick up new version'
+  # Fork a new task since this main process is started by RightLink and we are restarting it.
+  Start-Process Powershell -ArgumentList "-Command & { $upgradeFunction upgradeRightLink ${currentVersion} ${desiredVersion} }"
+} else {
+  Write-Output "Updated version does not appear to be desired version: ${newVersion}"
```

Comment:
Exit code was added, not sure why it isn't updating this as a comment on an outdated diff.